pyfr/plugins/fwh.py

```diff
@@ -0,0 +1,208 @@
+from collections import namedtuple
+
+import numpy as np
+
+from pyfr.mpiutil import get_comm_rank_root, mpi
+from pyfr.nputil import npeval
+from pyfr.plugins.base import BaseSolnPlugin, init_csv, SurfaceMixin
+
+
+FWHSurfParams = namedtuple('FWHSurfParams', ('eidxs', 'm0', 'norm', 'area',
+                                             'r_tilde_vec', 'r_star_inv',
+                                             'r_star_tilde_vec'))
+
+
+class FWHPlugin(SurfaceMixin, BaseSolnPlugin):
+    name = 'fwh'
+    systems = ['ac-euler', 'ac-navier-stokes', 'euler', 'navier-stokes']
+    formulations = ['dual', 'std']
+    dimensions = [2, 3]
+
+    def __init__(self, intg, cfgsect, suffix=None, *args, **kwargs):
+        super().__init__(intg, cfgsect, suffix)
+        comm, rank, root = get_comm_rank_root()
+
+        self.elementscls = intg.system.elementscls
+
+        self.tstart = self.cfg.getfloat(cfgsect, 'tstart', 0.0)
+        self.t_last = -np.inf
+        self.dt = self.cfg.getfloat(cfgsect, 'dt')
+        self.obsv_pts = np.array(self.cfg.getliteral(self.cfgsect,
+                                                     'observer-pts'))
+        self.nobvs = len(self.obsv_pts)
+
+        # Initialise data file
+        ndims = self.ndims
+        if rank == root:
+            header = ','.join(['t', 'x', 'y', 'z'][:self.ndims + 1] + ['mag'])
+            self.outf = init_csv(self.cfg, cfgsect, header)
+            self.outf.flush()
+
+        # Far field conditions
+        self.incomp = intg.system.name in {'ac-euler', 'ac-navier-stokes'}
+        privarmap = intg.system.elementscls.privarmap[ndims]
+        self._vidx = [x in 'uvw' for x in privarmap]
+        self._pidx = privarmap.index('p')
+        self.consts = self.cfg.items_as('constants', float)
+
+        self.qinf = {k: npeval(self.cfg.getexpr(cfgsect, k), self.consts)
+                     for k in privarmap}
+        self.uinf = np.array([[self.qinf[k]] for k in 'uvw'[:ndims]])
+
+        if self.incomp:
+            self.qinf['rho'] = self.cfg.getfloat(cfgsect, 'rho')
+            self.qinf['c'] = self.cfg.getfloat(cfgsect, 'c')
+        else:
+            gamma = self.consts['gamma']
+            self.qinf['c'] = (gamma * self.qinf['p'] / self.qinf['rho'])**0.5
+
+        self.qinf['M'] = np.array([self.qinf[k]/self.qinf['c'] for k in
+                                   'uvw'[:ndims]])
+        self.Minf = np.linalg.norm(self.qinf['M'])
+        if self.Minf >= 1:
+            raise ValueError("FWH farfield Mach number greater than 1.")
```

Comment:
Single quotes, no full stop.  Can use := for the assignment if you like.

Reply:
I don't think you can use `:=` for an attribute of a class.